### ISSUE TITLE
evaluate speedup via mixed-precision

### ISSUE THREAD

**Issue by ArneBinder** [ArneBinder/pie-document-level#358]
Should work out of the box via https://github.com/ArneBinder/pytorch-ie/pull/442 (included in [pytorch-ie>=0.31.5](https://github.com/ArneBinder/pytorch-ie/releases/tag/v0.31.5)).

**Comment by ArneBinder** [comment]
# Evaluate Joint Model
 - base setup: https://github.com/ArneBinder/pie-document-level/issues/334#issuecomment-2613232344 
 - on the test set
 - batch_size=1 because it is the fastest according to https://github.com/ArneBinder/pie-document-level/issues/334#issuecomment-2613304742

## baseline:

inference command:
```
python src/predict.py \
dataset=sciarg_prepared_with_fixed_validation_and_more_train \
dataset_split=test \
+pipeline.device=0 \
+pipeline.batch_size=1 \
+pipeline.half_precision_model=false \
model_name_or_path=models/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-53 \
ckpt_path=logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/0/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/1/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/2/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/3/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/4/checkpoints/last.ckpt \
--multirun
```
result location: `logs/prediction/multiruns/default/2025-01-24_22-26-23/job_return_value.aggregated.md`

|                 |   count |    mean |     std |
|:----------------|--------:|--------:|--------:|
| prediction_time |       5 | 131.714 | 4.10725 |


## Mixed precision

inference command:
```
python src/predict.py \
dataset=sciarg_prepared_with_fixed_validation_and_more_train \
dataset_split=test \
+pipeline.device=0 \
+pipeline.batch_size=1 \
+pipeline.half_precision_model=true \
model_name_or_path=models/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-53 \
ckpt_path=logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/0/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/1/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/2/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/3/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/4/checkpoints/last.ckpt \
--multirun
```
result location: `logs/prediction/multiruns/default/2025-01-24_22-42-02/job_return_value.aggregated.md`

|                 |   count |    mean |     std |
|:----------------|--------:|--------:|--------:|
| prediction_time |       5 | 130.563 | 3.90689 |

**OUTCOME: The result is identical, did this actually work at all?** 

## Mixed precision and bigger batch size
- maybe there is no effect on too small batch_size? we try a bigger one
inference command:
```
python src/predict.py \
dataset=sciarg_prepared_with_fixed_validation_and_more_train \
dataset_split=test \
+pipeline.device=0 \
+pipeline.batch_size=16 \
+pipeline.half_precision_model=true \
model_name_or_path=models/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-53 \
ckpt_path=logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/0/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/1/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/2/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/3/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/4/checkpoints/last.ckpt \
--multirun
```
result location: `logs/prediction/multiruns/default/2025-01-24_23-45-32/job_return_value.aggregated.md`

|                 |   count |    mean |     std |
|:----------------|--------:|--------:|--------:|
| prediction_time |       5 | 253.062 | 51.3163 |


## Mixed precision and bigger batch size - baseline

inference command:
```
python src/predict.py \
dataset=sciarg_prepared_with_fixed_validation_and_more_train \
dataset_split=test \
+pipeline.device=0 \
+pipeline.batch_size=16 \
+pipeline.half_precision_model=false \
model_name_or_path=models/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-53 \
ckpt_path=logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/0/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/1/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/2/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/3/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/4/checkpoints/last.ckpt \
--multirun
```
result location: `logs/prediction/multiruns/default/2025-02-06_11-54-48/job_return_value.aggregated.json`

|                 |   count |    mean |    std |
|:----------------|--------:|--------:|-------:|
| prediction_time |       5 | 254.431 | 51.396 |

## Mixed Precision with pie-modules PR
- maybe this  is necessary? https://github.com/ArneBinder/pie-modules/pull/149
```
python src/predict.py \
dataset=sciarg_prepared_with_fixed_validation_and_more_train \
dataset_split=test \
+pipeline.device=0 \
+pipeline.batch_size=1 \
+pipeline.half_precision_model=true \
model_name_or_path=models/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-53 \
ckpt_path=logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/0/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/1/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/2/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/3/checkpoints/last.ckpt,logs/training/multiruns/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-52/4/checkpoints/last.ckpt \
--multirun
```
result location: `logs/prediction/multiruns/default/2025-01-25_00-57-13/job_return_value.aggregated.md`

|                 |   count |    mean |     std |
|:----------------|--------:|--------:|--------:|
| prediction_time |       5 | 131.497 | 4.27227 |

**Comment by ArneBinder** [comment]
# Evaluate Pipelined Model
 - base setup: https://github.com/ArneBinder/pie-document-level/issues/355#issuecomment-2609779724 (this is on the test set)

## baseline

from https://github.com/ArneBinder/pie-document-level/issues/355#issuecomment-2609779724

|                 |   count |    mean |     std |
|:----------------|--------:|--------:|--------:|
| prediction_time |      25 | 230.578 | 11.5275 |

## mixed precision for RE model

```
python src/predict.py \
dataset=sciarg_prepared \
pipeline=ner_re_pipeline \
pipeline.ner_model_path=models/dataset-sciarg/task-adus/v0.4/2025-01-20_02-30-35,models/dataset-sciarg/task-adus/v0.4/2025-01-20_04-11-02,models/dataset-sciarg/task-adus/v0.4/2025-01-20_05-50-00,models/dataset-sciarg/task-adus/v0.4/2025-01-20_07-29-42,models/dataset-sciarg/task-adus/v0.4/2025-01-20_09-09-11 \
pipeline.re_model_path=models/dataset-sciarg/task-relations/v0.4/2025-01-22_12-44-51,models/dataset-sciarg/task-relations/v0.4/2025-01-22_16-40-45,models/dataset-sciarg/task-relations/v0.4/2025-01-22_20-36-23,models/dataset-sciarg/task-relations/v0.4/2025-01-22_12-36-22,models/dataset-sciarg/task-relations/v0.4/2025-01-22_16-35-41 \
pipeline.entity_layer=labeled_spans \
pipeline.relation_layer=binary_relations \
++pipeline.device=0 \
++pipeline.ner_pipeline.batch_size=256 \
++pipeline.re_pipeline.batch_size=64 \
+pipeline.re_pipeline.half_precision_model=true \
--multirun
```
result location: `logs/prediction/multiruns/default/2025-01-24_22-55-17/job_return_value.aggregated.md`

|                 |   count |    mean |     std |
|:----------------|--------:|--------:|--------:|
| prediction_time |      25 | 79.4275 | 3.95699

**Comment by ArneBinder** [comment]
# Evaluate Joint Model - *with* pie-modules fix, but *not* from checkpoint
- loading a checkpoint seems to revert the model to full precision
   - EDIT: see #375 
- requires https://github.com/ArneBinder/pie-modules/pull/154, i.e. branch `pointer_network/half_precision`:
  - `pip uninstall pie-modules && pip install git+https://github.com/ArneBinder/pie-modules.git@pointer_network/half_precision`
- just one run for now
- inference command:
  ```
  python src/predict.py \
  dataset=sciarg_prepared_with_fixed_validation_and_more_train \
  dataset_split=test \
  +pipeline.device=0 \
  +pipeline.batch_size=1 \
  +pipeline.half_precision_model=true \
  +pipeline.half_precision_ops=false \
  model_name_or_path=models/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-53
  ``` 
- result location: `logs/prediction/runs/default/2025-02-06_19-00-10/job_return_value.json`
- result: "prediction_time": 133.72791209816933

## larger batch size

```
python src/predict.py \
dataset=sciarg_prepared_with_fixed_validation_and_more_train \
dataset_split=test \
+pipeline.device=0 \
+pipeline.batch_size=16 \
+pipeline.half_precision_model=true \
+pipeline.half_precision_ops=false \
model_name_or_path=models/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-53
``` 
- result location: `logs/prediction/runs/default/2025-02-06_19-05-30/job_return_value.json`
- result: "prediction_time": 212.41, this is smaller than mean baseline 254.431 (but the stddev is 50!)
- OUTCOME: seems to have an effect, but not great

## even larger batch size

```
python src/predict.py \
dataset=sciarg_prepared_with_fixed_validation_and_more_train \
dataset_split=test \
+pipeline.device=0 \
+pipeline.batch_size=64 \
+pipeline.half_precision_model=true \
+pipeline.half_precision_ops=false \
model_name_or_path=models/dataset-sciarg/task-ner_re/v0.4/2025-01-09_01-50-53
``` 
- result location: `logs/prediction/runs/default/2025-02-06_19-18-14/job_return_value.json`
- result: "prediction_time": 268.67

## OUTCOME: STILL NOT VERY CLEAR IF THIS IS WORKING AS EXPECTED
